crates/esplora/src/async_ext.rs

```diff
@@ -259,7 +259,12 @@ impl EsploraAsyncExt for esplora_client::AsyncClient {
                     }
                 }
 
-                if last_index > last_active_index.map(|i| i.saturating_add(stop_gap as u32)) {
+                let last_index = last_index.expect("Must be set since handles wasn't empty.");
+                if last_index
+                    > last_active_index
+                        .unwrap_or(0)
```

Comment:
Is this right? A last_active_index of `None` and a `last_active_index` of 0 are not the same, they differ by 1 spk seen. Maybe this should be `unwrap_or(-1)`?

Reply:
Heh, true. Can't use -1 though as it's unsigned. So it looks like i need to do something like this to be accurate (and still clear):
```diff
diff --git a/crates/esplora/src/blocking_ext.rs b/crates/esplora/src/blocking_ext.rs
index 7ee4625..bde24f8 100644
--- a/crates/esplora/src/blocking_ext.rs
+++ b/crates/esplora/src/blocking_ext.rs
@@ -251,11 +251,12 @@ impl EsploraExt for esplora_client::BlockingClient {
                 }
 
                 let last_index = last_index.expect("Must be set since handles wasn't empty.");
-                if last_index
-                    > last_active_index
-                        .unwrap_or(0)
-                        .saturating_add(stop_gap as u32)
-                {
+                let past_gap_limit = if let Some(i) = last_active_index {
+                    last_index > i.saturating_add(stop_gap as u32)
+                } else {
+                    last_index >= stop_gap as u32
+                };
+                if past_gap_limit {
                     break;
                 }
             }
```

However looks like this uncovered a crash now..
```
running 1 test
test test_update_tx_graph_gap_limit ... FAILED

failures:

---- test_update_tx_graph_gap_limit stdout ----
thread 'test_update_tx_graph_gap_limit' panicked at crates/esplora/tests/blocking_ext.rs:183:47:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: rust_begin_unwind
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:595:5
   1: core::panicking::panic_fmt
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/panicking.rs:67:14
   2: core::panicking::panic
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/panicking.rs:117:5
   3: core::option::Option<T>::unwrap
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/option.rs:935:21
   4: blocking_ext::test_update_tx_graph_gap_limit
             at ./tests/blocking_ext.rs:183:16
   5: blocking_ext::test_update_tx_graph_gap_limit::{{closure}}
             at ./tests/blocking_ext.rs:120:44
   6: core::ops::function::FnOnce::call_once
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/ops/function.rs:250:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    test_update_tx_graph_gap_limit
```

Reply:
Interesting! I'm looking into it. I want to merge your fix, but first I want to make sure that we're not introducing a bigger bug.

Reply:
Of course! Thanks for looking into it.

Reply:
The failure you see is in the tests, because you removed one off-by-one (the `unwrap_or(0)`), but there's another off-by-one in the code posted.
I started by updating the tests following my comment https://github.com/bitcoindevkit/bdk/pull/1225#discussion_r1402460854, so that we don't find the tx to spk3 with stop_gap = 3, but we do find it with stop_gap = 4; and we don't find the tx to spk9 with a stop_gap of 5, but we find it with a stop_gap of 6:
```diff
diff --git a/crates/esplora/tests/async_ext.rs b/crates/esplora/tests/async_ext.rs
index ec79cd07..409daf04 100644
--- a/crates/esplora/tests/async_ext.rs
+++ b/crates/esplora/tests/async_ext.rs
@@ -164,7 +164,7 @@ pub async fn test_async_update_tx_graph_gap_limit() -> anyhow::Result<()> {
         sleep(Duration::from_millis(10))
     }
 
-    // A scan with a gap limit of 1 won't find the transaction, but a scan with a gap limit of 2
+    // A scan with a gap limit of 3 won't find the transaction, but a scan with a gap limit of 4
     // will.
     let (graph_update, active_indices) = env
         .client
@@ -172,7 +172,7 @@ pub async fn test_async_update_tx_graph_gap_limit() -> anyhow::Result<()> {
             keychains.clone(),
             vec![].into_iter(),
             vec![].into_iter(),
-            1,
+            3,
             1,
         )
         .await?;
@@ -184,7 +184,7 @@ pub async fn test_async_update_tx_graph_gap_limit() -> anyhow::Result<()> {
             keychains.clone(),
             vec![].into_iter(),
             vec![].into_iter(),
-            2,
+            4,
             1,
         )
         .await?;
@@ -207,7 +207,7 @@ pub async fn test_async_update_tx_graph_gap_limit() -> anyhow::Result<()> {
         sleep(Duration::from_millis(10))
     }
 
-    // A scan with gap limit 4 won't find the second transaction, but a scan with gap limit 5 will.
+    // A scan with gap limit 5 won't find the second transaction, but a scan with gap limit 6 will.
     // The last active indice won't be updated in the first case but will in the second one.
     let (graph_update, active_indices) = env
         .client
@@ -215,7 +215,7 @@ pub async fn test_async_update_tx_graph_gap_limit() -> anyhow::Result<()> {
             keychains.clone(),
             vec![].into_iter(),
             vec![].into_iter(),
-            4,
+            5,
             1,
         )
         .await?;
@@ -225,7 +225,7 @@ pub async fn test_async_update_tx_graph_gap_limit() -> anyhow::Result<()> {
     assert_eq!(active_indices[&0], 3);
     let (graph_update, active_indices) = env
         .client
-        .scan_txs_with_keychains(keychains, vec![].into_iter(), vec![].into_iter(), 5, 1)
+        .scan_txs_with_keychains(keychains, vec![].into_iter(), vec![].into_iter(), 6, 1)
         .await?;
     let txs: HashSet<_> = graph_update.full_txs().map(|tx| tx.txid).collect();
     assert_eq!(txs.len(), 2);
```

Then, the code should be modified to:
```
+                let past_gap_limit = if let Some(i) = last_active_index {
+                    last_index >= i.saturating_add(stop_gap as u32)
+                } else {
+                    // If the last active is None, last_index + 1 is the number of spks we searched
+                    // If we searched `stop_gap` or more spks without finding a tx, then we're past
+                    // the gap_limit
+                    last_index + 1 >= stop_gap as u32
+                };
```

Using `>=` instead of `>` in the first arm, and `last_index + 1` in the second arm.

Does this sound fair?

Reply:
Partially. You are right for the case where there is a last active index. I had the logic off by one, i was considering the next index. But i think you are wrong for the case where there is no last active index. For instance take a gap limit of 1. Your code would make us break if we just scanned index 0 unsuccessfully, although we should also scan index 1. (Only scanning index 0 is gap limit 0.)

Reply:
No actually i think you are also wrong in the first case. Take a gap limit of 1. If our last active index is 4 and we just scanned index 5 unsuccessfully we should also scan index 6. (Only scanning index 5 is gap limit 0.)

So i think my patch was in fact correct https://github.com/bitcoindevkit/bdk/pull/1225#discussion_r1402485388. :)

Reply:
Mhh, let's start from the tests. In the first case, what's the min stop gap that must find the tx? Is it 3, or is it 4? 

I really want to have a clear definition in my mind of gap limit before reasoning about this :)